Git fix read wyrm

gcater opened the pull request:
Added components for auto form, added functionality for TRPC-can add books with chapters and section, allow for dynamic routes depeding on a book's id/chapter/section.


<img width="1619" alt="Screenshot 2024-04-03 at 5 29 59 PM" src="https://github.com/gcater/bookwyrm/assets/109175497/31e09872-7bd0-4b75-ac77-a01ed03bb0fc">
<img width="1594" alt="Screenshot 2024-04-03 at 5 30 36 PM" src="https://github.com/gcater/bookwyrm/assets/109175497/2fbef583-d0fe-4a44-a6db-19a785f93f5a">
<img width="1610" alt="Screenshot 2024-04-03 at 5 30 43 PM" src="https://github.com/gcater/bookwyrm/assets/109175497/3b3c1fa1-dc9d-4834-a3a0-975491c1bfb7">
<img width="1559" alt="Screenshot 2024-04-03 at 5 30 52 PM" src="https://github.com/gcater/bookwyrm/assets/109175497/04c2457a-9c34-4c72-aa66-20c02d96d1b0">

Evanc123 commented on src/server/api/routers/book.ts:
```diff
@@ -1,27 +1,203 @@
+import { createTRPCRouter, protectedProcedure } from "src/server/api/trpc";
 
-import { createTRPCRouter, protectedProcedure } from "~/server/api/trpc";
+export interface Section {
```
think you shouldn't need these. can be inferred by either `routerOutputs` or from Prisma.getPayload, probably the former is better.

Evanc123 commented on src/server/api/routers/book.ts:
```diff
@@ -1,27 +1,203 @@
-          createdById: ctx.session.user.id,
+          createdBy: { connect: { id: ctx.session.user.id } },
+          chapters: {
+            create: input.chapters.map((chapter: Chapter) => ({
```
yeah this is not needed, will be inferred
```suggestion
            create: input.chapters.map((chapter) => ({
```

Evanc123 commented on src/server/api/routers/book.ts:
```diff
@@ -1,27 +1,203 @@
           title: input.title,
           author: input.author,
-          createdById: ctx.session.user.id,
+          createdBy: { connect: { id: ctx.session.user.id } },
```
lets just do createdById and forgo the connect (which will happen automatically)

Evanc123 commented on src/server/api/routers/book.ts:
```diff
@@ -1,27 +1,203 @@
 
-  getAll: protectedProcedure.query(({ ctx }) => {
-    return ctx.db.book.findMany({
+  delete: protectedProcedure.input(z.string()).mutation(async ({ ctx, input }) => {
```
lets do a named string here, call it `bookIdToDelete` or just `bookId` (latter is probably better)

Evanc123 commented on src/pages/book/networkstate.tsx:
```diff
@@ -0,0 +1,38 @@
+        <div className="flex w-full flex-col rounded-xl border border-gray-300">
+          <div className="border-b border-gray-300 last:border-none">
+            {/* <Link href="/index">
+              <Link href="/index"> */}
```
let's try to cleanup comments in general, helps with sanity / sanitization

Evanc123 replied:
(that is, delete uncommented code blocks pre-pr)

Evanc123 commented on src/pages/book/[bookid]/index.tsx:
```diff
@@ -0,0 +1,51 @@
+
+export default function AddChapterPage() {
+  const router = useRouter();
+  const { bookid } = router.query; // `bookid` matches the dynamic segment name
```
`bookid` should be camelcased, `bookId`

Evanc123 commented on src/pages/book/[bookid]/index.tsx:
```diff
@@ -0,0 +1,51 @@
+"use client";
```
you rpobably don't need this

Evanc123 commented on src/pages/book/[bookid]/index.tsx:
```diff
@@ -0,0 +1,51 @@
+import BookForm from "~/components/BookForm";
+import BookRender from "~/components/BookRender";
+import AutoBookForm from "~/components/AutoBookForm";
+import AutoChapterForm from "~/components/AutoChapterForm";
```
You can just call these `BookForm` and `ChapterForm`

Evanc123 commented on src/pages/book/[bookid]/index.tsx:
```diff
@@ -0,0 +1,51 @@
+  return (
+    <>
+      <Head>
+        <title>Create T3 App</title>
```
change

Evanc123 commented on src/pages/book/[bookid]/addChapter.tsx:
```diff
@@ -0,0 +1,51 @@
+        </div>
+        <div className="flex w-full">
+          <div className="w-1/2">
+            <AutoChapterForm bookId={bookid as string} />
```
put the `as string` after the router.query instead of here. also add an if statement, like 

if (!bookId) {
return <div>No book found</div>
} 

or something liek that

Evanc123 commented on src/lib/utils.ts:
```diff
@@ -0,0 +1,6 @@
+import { type ClassValue, clsx } from "clsx"
+import { twMerge } from "tailwind-merge"
```
next time when adding a big library like shadcn, let's split it up into two prs so they are more distinct / easier to review. 

best to have Prs be distinct cuts, easier to revert that way and way easier to review

Evanc123 commented on src/components/bookRender.tsx:
```diff
@@ -0,0 +1,99 @@
+
+  return (
+    // <div className="m-0 flex flex-grow flex-col p-0 sm:ml-[90px] sm:pl-[0 px] sm:pr-[100px]">
+    //   <div className="hidden w-full flex-col border-b border-gray-300 pb-6 sm:flex">
```
clean this up

Evanc123 commented on src/components/bookRender.tsx:
```diff
@@ -0,0 +1,99 @@
+type CardProps = React.ComponentProps<typeof Card>;
+
+const BookRender = (): JSX.Element => {
+  return ShadCard({});
```
whats this?

Evanc123 commented on src/components/bookRender.tsx:
```diff
@@ -0,0 +1,99 @@
+  return ShadCard({});
+};
+
+export function ShadCard({ className, ...props }: CardProps) {
```
not sure if we want this to be ShadCard. Open to it but i don't understand whats going on

Evanc123 commented on src/components/bookForm.tsx:
```diff
@@ -0,0 +1,152 @@
+
+type FormValues = z.infer<typeof formSchema>;
+// This will return the book form JSX Element. It will allow a user to type in the author, title, and add chapters. Each chapter will have a title and can have multiple sections. Sections will have a title and text content.
+const BookForm = (): JSX.Element => {
```
i don't think we need any of this

Evanc123 commented on src/components/bookForm.tsx:
```diff
@@ -0,0 +1,152 @@
+  };
+
+  return (
+    <form onSubmit={handleSubmit(onSubmit)}>
```
the point of using autofrom is that we don't need to do this by hand

Evanc123 commented on src/components/AutoSectionForm.tsx:
```diff
@@ -0,0 +1,97 @@
+    //   // Pass the schema to the form
+    //   formSchema={formSchema}
+    // >
+    //   <AutoFormSubmit>Send now</AutoFormSubmit>
```
clean

Evanc123 commented on src/components/AutoSectionForm.tsx:
```diff
@@ -0,0 +1,97 @@
+  
+});
+
+const MyAutoForm = ({bookId, chapterId}: AutoSectionFormProps): JSX.Element => {
```
let's keep names consistent. `ChapterForm` or something like that

Evanc123 commented on src/components/AutoSectionForm.tsx:
```diff
@@ -0,0 +1,97 @@
+
+}
+
+const formSchema = z.object({
```
let's give these schemas semantic meaning

Evanc123 commented on src/components/AutoChapterForm.tsx:
```diff
@@ -0,0 +1,111 @@
+      All children passed to the form will be rendered below the form.
+      */}
+            <p className="text-sm text-gray-500">
+              By submitting this form, you agree to our{" "}
```
clean this up

Evanc123 commented on src/components/AutoChapterForm.tsx:
```diff
@@ -0,0 +1,111 @@
+        sections: [], // Assuming sections can be an empty array for now
+      },
+      bookId,
+      //bookId: parseInt(bookId, 10), // Convert bookId from string to number, ensure bookId is not null before this operation
```
clean this up

Evanc123 commented on src/components/AutoChapterForm.tsx:
```diff
@@ -0,0 +1,111 @@
+  bookId: string;
+}
+
+const MyAutoForm = ({ bookId }: AutoChapterFormProps): JSX.Element => {
```
lets name this something semantic

Evanc123 commented on src/components/AutoChapterForm.tsx:
```diff
@@ -0,0 +1,111 @@
+import { Button } from "@/components/ui/button";
+import { api } from "~/utils/api";
+import type { Book } from "~/server/api/routers/book";
+import React, { useState } from "react"; // Import useState
```
remove comment

Evanc123 commented on src/components/AutoBookForm.tsx:
```diff
@@ -0,0 +1,126 @@
+            <CardContent>
+              <Button
+                onClick={() => {
+                  if (window.confirm("Are you sure you want to delete this book?")) {
```
lets use a shadcn `Popover` instead of window.confirm

Evanc123 reviewed (changes requested):
Lot's of little things! let's get another pass going

gcater commented:
AutoBookForm, AutoChapterForm, and AutoSectionForm components have been renamed to BookForm, ChapterForm, and SectionForm. They no longer require explicit interface imports for Book/Chapter/Section from book.ts. These types are now inferred from the routers. Cleaned up comments and content that was not necessary. 

The original BookForm component that did not use the ShadCN Autoform has been deleted.

The BookRender.tsx component has been renamed BookRenderer

In book.ts:
Renamed functions to specify what they do create -> createBook, delete->deleteBook. 
Changed the createBook field data from createdBy: { connect: { id: ctx.session.user.id } } to createdById: ctx.session.user.id. createBook now does not worry about chapters/sections. 
addChapter does not worry about sections.
The deleteBook function now takes in "z.object({ bookId: z.string() }" as input. 

src/pages/book/[bookid]/index.tsx does and  src/pages/book/networkstate.tsx were deleted as they went unused.

gcater commented:
Changed the logic in BookForm.tsx to handle deletion. Previously it would cause a react error. It now features a ShadCN popover.

gcater commented:
Added type checks and null checks for bookId in addChapter.tsx and bookId/sectionId in addSection.tsx.

Renamed the formSchemas in BookForm, ChapterForm, and SectionForm to bookInputs, chapterInputs, and sectionInputs.

Got rid of the functions shadCard in BookRenderer. I think I was using it because I was unsure how card properties worked. It was unnecessary.

I believe that's covers all of the requested changes.

Evanc123 commented on src/components/BookForm.tsx:
```diff
@@ -0,0 +1,104 @@
+  PopoverTrigger,
+} from "@/components/ui/popover";
+
+const bookInputs = z.object({
```
zod schema's are always postfixed with 'schema`, so in this case we would call it 
```suggestion
const BookInputSchema = z.object({
```

Evanc123 commented on src/components/ChapterForm.tsx:
```diff
@@ -0,0 +1,67 @@
+}
+
+const ChapterForm = ({ bookId }: ChapterFormProps): JSX.Element => {
+  const { mutate, data, isSuccess } = api.book.addChapter.useMutation({
```
do we need `data` here?

Evanc123 replied:
as best practice, we should make sure to name our vars, which doesn't really matter for components this small but for larger components keeps everything in check  
```suggestion
  const { mutate: addChapterMutation, data, isSuccess: isAddChapterMutationSuccessful } = api.book.addChapter.useMutation({
```

Evanc123 commented on src/components/bookRenderer.tsx:
```diff
@@ -0,0 +1,57 @@
+} from "@/components/ui/card";
+type CardProps = React.ComponentProps<typeof Card>;
+
+const BookRenderer = ({ className, ...props }: CardProps) => {
```
so this doesn't really need to take in `CardProps` right? Nor does it need `className`? What does the bookRenderer really need to take in? that should be the component's props.

Evanc123 commented on src/pages/book/[bookid]/[chapterid]/addSection.tsx:
```diff
@@ -0,0 +1,36 @@
+    <>
+      <Head>
+        <title>Add Section</title>
+        <meta name="description" content="Generated by create-t3-app" />
```
fix

Evanc123 commented on src/pages/book/[bookid]/[chapterid]/addSection.tsx:
```diff
@@ -0,0 +1,36 @@
+import SectionForm from "~/components/SectionForm";
+export default function AddSectionPage() {
+  const router = useRouter();
+  const bookId = router.query.bookid as string; // Moved and type-checked
```
don't need the comment here

Evanc123 commented on src/pages/book/[bookid]/[chapterid]/addSection.tsx:
```diff
@@ -0,0 +1,36 @@
+  const bookId = router.query.bookid as string; // Moved and type-checked
+  const chapterId = router.query.chapterid as string; // Moved and type-checked
+
+  if (!bookId || !chapterId) return <p>No book or chapter found!</p>; // Added check for both bookId and chapterId
```
keep comments in the commit or the PR comments, code comments should only be semantic to the code, not about the meta-change

Evanc123 commented on src/server/api/routers/book.ts:
```diff
@@ -1,27 +1,141 @@
+      z.object({
+        bookId: z.string(),
+        chapterId: z.string(),
+        section: z.object({
```
you've defined this zod object somehwere, right? if you replace it with `SectionFormSchema` (whereever that is defined), then it will be easier to maintain once you do make a future change to `SectionSchema` (which we will undoubtedly change)

Evanc123 reviewed (approved):
Approved pending the listed changes
